certbot/plugins/dns_common.py

```diff
@@ -29,13 +54,68 @@ def __init__(self, config, name):
 
         self._attempt_cleanup = False
 
+    @classmethod
+    def inject_parser_options(cls, parser, name):
+        """Set up argument parsing for this DNS plugin.
+
+        This is called by the plugin management framework to set up
+        the argument parser.
+
+        :param ArgumentParser parser: (Almost) top-level CLI parser.
+        :param str name: Unique plugin name.
+        """
+
+        # The inherited method will call add_parser_arguments to set up
+        # the parser of each argument.
+
+        super(DNSAuthenticator, cls).inject_parser_options(parser, name)
+
+        # Create an additional entry in the parser namespace for the
+        # override-challenge map. This is an initiallyt empty dict associating
+        # each requested certificate domain with the corresponding challenge
+        # override (see help for --dns-<plugin>-override-challenge).
+
+        # Use set_default to provide this dict as an entry in the namespace
+        # object returned by the parser. Note that set_default uses keyword
+        # arguments to determine the name of the entry to be set, and
+        # here the name is dynamically computed (depending on the plugin
+        # name), which is why we need to use ** syntax.
+
+        challenge_map_opt = common.dest_namespace(name) + "override_challenge_map"
+        parser.set_defaults(**{challenge_map_opt: {}})  # pylint: disable=star-args
+
+
     @classmethod
     def add_parser_arguments(cls, add, default_propagation_seconds=10):  # pylint: disable=arguments-differ
         add('propagation-seconds',
             default=default_propagation_seconds,
             type=int,
-            help='The number of seconds to wait for DNS to propagate before asking the ACME server '
-                 'to verify the DNS record.')
+            help='The number of seconds to wait for DNS to propagate before '
+                 'asking the ACME server to verify the DNS record.')
+        add('override-challenge',
+            action=OverrideChallengeAction,
+            default="{acme}",
+            help='Override default challenge for following domains on the '
+                 'command line. Validation depends on a CNAME having been '
+                 'provisioned at the standard location '
+                 '(_acme-challenge.<domain>), pointing to the alternate '
+                 'location. The specified value may include "{domain}" '
+                 'and "{acme}" format strings, which will respectively '
+                 'expand to the domain being requested, and the corresponding '
+                 'standard ACME challenge location. For instance, when '
+                 'requesting "example.com", {domain} is "example.com", and '
+                 '{acme} is "_acme-challenge.example.com".')
```

Comment:
I see the concern about the help text now. Not only is how to use this feature with CNAMEs somewhat complex, but this flag works like `--webroot-path` and includes a templating system.

I personally think we can't just say something like:
```
This flag is complex. Go read <url> to learn how to use it.
```
I think we should try to:

1. Provide a succinct but sufficient explanation about what this flag does for an experienced, technical user.
2. Maybe provide a link at the bottom to learn more.

@schoen, if you agree with me, are you interested in trying to propose text for (1) here? I think you would do a much better job at this than me.

Reply:
It is also what I proposed to @joohoi on the last discussion we had: just put a link to the web documentation in the CLI help, because I do not see how to make a short summary of this feature that would be understandable.

The main concern in the web documentation on which we are agree also, is that the documentation must not give to users a false feeling of security just because they use a delegate CNAME. There is more to do to build a really secured way that protect DNS credentials, and some work and study need to be done on the users side.

Reply:
I think I'd personally rather have long help text about this flag than no help text other than a link. What harm is there in including an explanation of whatever length is necessary?

I think this help text should focus on what the flag **does** and we can put a link if we want explaining who, what, where, when, why, and how someone might want to use this feature.

Do people disagree? If so, why?

Reply:
Oh I misunderstood your last response. You want a description. Still, I do not see how to write something moderately short, as it involves a lot of context to understand what is its usage. Even by limiting to what it is doing.

Reply:
I think a native english speaker should handle the job. I would be not sufficiently clear, or too verbose. @schoen or @bmw, one of you would be willing to make a proposition?